### PR TITLE
WIP: Implement 'ParallelImageCapture' for samd51

### PR DESCRIPTION
**Pull request by jepler**
This work-in-progress code, together with a WIP CircuitPython library, can successfully capture small (80x60) frames from an OV7670 camera on a Metro M4 GrandCentral.

TODO:
- [x] a name everyone can agree on
- [x] capture using DMA, rather than having interrupts disabled for 30ms

cap

### CODE REVIEW

File: shared-bindings/imagecapture/ParallelImageCapture.c [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+//|     """Capture image frames from a camera with parallel data interface"""
+//|
+//|     def __init__(
+//|         self,
+//|         *,
+//|         data0: microcontroller.Pin,
+//|         data_count: int=8,
+//|         clock: microcontroller.Pin,
+//|         vsync: Optional[microcontroller.Pin],
+//|         hsync: Optional[microcontroller.Pin],
```
Comment (tannewt) [inline shared-bindings/imagecapture/ParallelImageCapture.c:46]: Should this be HREF instead of HSYNC?

File: locale/circuitpython.pot [hunk 2]
```diff
@@ -1705,6 +1711,11 @@ msgstr ""
 msgid "Oversample must be multiple of 8."
 msgstr ""
 
+#: ports/atmel-samd/common-hal/imagecapture/ParallelImageCapture.c
+#, c-format
+msgid "PCC_D%d in use"
```
Comment (tannewt) [inline locale/circuitpython.pot:1716]: Please make this a generic message. PCC is SAMD specific.

**Review by tannewt — COMMENTED** [review]
Thanks for the module rename. I like `imagecapture`. Just a couple other minor things.

I think you need per-param comments for the constructor too.

**Comment by jepler** [comment]
@tannewt this is ready for a fresh review.

Main question **I** have right now is, I used "audio_dma" channels. As far as I can see this will work just fine (but didn't test): with 3 channels you can play stereo audio in the background and capture a framebuffer image in the foreground; we'd have to increase the channel count in the future when we enable this or audio capture to also happen in the background.  That said, should the APIs `audio_dma_allocate_channel()` and `audio_dma_free_channel` be renamed?  Be moved to different locations?  Or can it be left as-is?

File: shared-bindings/displayio/ColorConverter.c [hunk 3]
```diff
@@ -49,18 +50,19 @@
 // TODO(tannewt): Add support for other color formats.
```
Comment (tannewt) [inline shared-bindings/displayio/ColorConverter.c:50]: Delete this and add a docstring for it.

File: shared-bindings/displayio/__init__.c [hunk 4]
```diff
@@ -65,10 +65,48 @@ STATIC mp_obj_t displayio_release_displays(void) {
...
+MAKE_ENUM_VALUE(displayio_colorspace_type, displayio_colorspace, RGB555_SWAPPED, DISPLAYIO_COLORSPACE_RGB555_SWAPPED);
+
+//| class Colorspace:
+//|     """The colorspace for a ColorConverter to operate in"""
+//|
+//|     RGB888: object
+//|     """The standard 24-bit colorspace"""
+//|
+//|     RGB565: object
+//|     """The standard 16-bit colorspace"""
```
Comment (tannewt) [inline shared-bindings/displayio/__init__.c:81]: Please elaborate on these comments so that no other reference info is needed.

Comment (tannewt) [inline shared-bindings/displayio/__init__.c:81]: Please elaborate on these comments so that no other reference info is needed.

**Review by tannewt — CHANGES_REQUESTED** [review]
I think it would be good to rename the audio_dma stuff. I'm trying to remember what makes those special so we could rename them that way. Is it that they can be triggered by events?

**Comment by jepler** [comment]
@tannewt this is ready for a fresh review assuming the CI passes.  Let me know if I need to merge main or rebase onto it.

**Review by tannewt — APPROVED** [review]
Code looks good. Thank you! You'll need to fix the CI though.

**Review by tannewt — APPROVED** [review]
Thank you!

**Comment by jepler** [comment]
@ladyada 
Test program for ov7670 + gcm4 + 1.8" tft shield: https://gist.github.com/645f7d052b63103111b011e507df38a1
Test program for ov7670 + rpi pico + 2.0" ST7789 breakout: https://gist.github.com/f357464e194a4ba85bbcd43378b1e189

I think the 2nd one needs the fix for the DIV2 resolution mode, https://github.com/adafruit/Adafruit_CircuitPython_OV7670/pull/4

**Pull request merged**